File: trees/1305. All Elements in Two Binary Search Trees/getAllElements.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inroderList = function(root) {
    let curr = root;
    let stack = [];
    let inorder = [];

    while (curr || stack.length > 0) {
        while (curr) {
            stack.push(curr);
            curr = curr.left;
        }
        curr = stack.pop();
        inorder.push(curr.val);
        curr = curr.right;
    }

    return inorder;
}

/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {number[]}
 */
var getAllElements = function(root1, root2) {
    let result = [];

    let inorder1 = inroderList(root1);
    let inorder2 = inroderList(root2);

    // Merge the two inorder lists
    let i = 0, j = 0;
    while (i < inorder1.length && j < inorder2.length) {
        if (inorder1[i] < inorder2[j]) {
            result.push(inorder1[i]);
            i++;
        } else {
            result.push(inorder2[j]);
            j++;
        }
    }

    // Add the remaining elements
    while (i < inorder1.length) {
        result.push(inorder1[i]);
        i++;
    }
    while (j < inorder2.length) {
        result.push(inorder2[j]);
        j++;
    }

    return result;
};

File: trees/1305. All Elements in Two Binary Search Trees/getAllElements.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def getAllElements(self, root1, root2):
        """
        Returns a list conatining all the values of both trees in ascending order,
        By doing an Inorder traversal of both the trees at the same time
        
        :type root1: Optional[TreeNode]
        :type root2: Optional[TreeNode]
        :rtype: List[int]
        """
        curr1 = root1
        curr2 = root2
        stack1 = []
        stack2 = []
        result = []

        while curr1 or curr2 or stack1 or stack2:
            while curr1:
                stack1.append(curr1)
                curr1 = curr1.left
            
            while curr2:
                stack2.append(curr2)
                curr2 = curr2.left
            
            if stack1 and not stack2:
                # If the tree1 has more node, but tree2 is done
                curr1 = stack1.pop()
                result.append(curr1.val)
                curr1 = curr1.right
            elif stack2 and not stack1:
                # If the tree2 has more node, but tree1 is done
                curr2 = stack2.pop()
                result.append(curr2.val)
                curr2 = curr2.right
            elif stack1 and stack2:
                # If both trees still have nodes
                if stack1[-1].val < stack2[-1].val:
                    # Continue with the inorder traversal of the first tree
                    # Keep the second tree where it is as we'll need to compare it's curr2 node with the next curr1
                    curr1 = stack1.pop()
                    result.append(curr1.val)
                    curr1 = curr1.right
                else:
                    curr2 = stack2.pop()
                    result.append(curr2.val)
                    curr2 = curr2.right
        
        return result